---
layout: lesson
---

{% include gh_variables.html %}

In a [previous lesson](https://cloud-span.github.io/01genomics/), you learned how to organise a sequencing project. You also learned about metadata best practises, and what information you will need, and how this is best stored. Finally, you learned how to connect to a cloud instance via a command line interface.

As you progress through this lesson, keep in mind that, even if you arenâ€™t going to be doing this same workflow in your research, you will be learning some very important lessons about using command-line bioinformatic tools. What you learn here will enable you to use a variety of bioinformatic tools with confidence and greatly enhance your research efficiency and productivity.

> ## Getting Started
>
> This lesson assumes no prior experience with the tools covered in the course.
> However, learners are expected to have some familiarity with biological concepts,
> including the concept of genomic variation within a population, 
as well as some basic experience using a command line interface to navigate file systems.  
>
> For a beginner-level overview of the command line, see the [Cloud-SPAN Prenomics pages](https://cloud-span.github.io/prenomics00-intro/).
> If you are unsure whether your skills/experience are sufficient, why not try our [self-assessment quiz](https://shiny.york.ac.uk/er13/prenomics-quiz/#section-why) to test your knowledge?
>
> This lesson is part of a course that uses data hosted on an Amazon Machine Instance (AMI). Course participants will be given
> information on how
> to log-in to the AMI during the course. Information on preparing for the course is provided on the [Cloud-SPAN Genomics setup page](https://cloud-span.github.io/01genomics/setup.html).
{: .prereq}
